src/shim/binary_spinning_sem.cc

```diff
@@ -42,3 +44,18 @@ void BinarySpinningSem::wait(bool spin) {
     _x.store(false, std::memory_order_release);
     pthread_spin_unlock(&_lock);
 }
+
+int BinarySpinningSem::trywait() {
+    int rv = sem_trywait(&_semaphore);
+    if (rv != 0) {
+        assert(rv == -1 && errno == EAGAIN);
+        return rv;
+    }
+
+    _spin_ctr = 0;
```

Comment:
Hm, maybe we should not make spin_ctr a member variable of `BinarySpinningSem`. We don't need the state outside of a `wait` call and it's just creating more, unnecessary work for us here.